crates/net/discv5/src/lib.rs

```diff
@@ -0,0 +1,580 @@
+//! Wrapper around [`discv5::Discv5`].
+
+use std::{fmt, net::IpAddr, sync::Arc, time::Duration};
+
+use ::enr::Enr;
+use alloy_rlp::Decodable;
+use derive_more::{Constructor, Deref, DerefMut};
+use enr::{uncompressed_to_compressed_id, EnrCombinedKeyWrapper};
+use futures::future::join_all;
+use itertools::Itertools;
+use reth_net_common::discovery::{HandleDiscovery, NodeFromExternalSource};
+use reth_primitives::{
+    bytes::{Bytes, BytesMut},
+    ForkId, NodeRecord, PeerId,
+};
+use secp256k1::SecretKey;
+use tokio::{sync::mpsc, task};
+use tracing::{debug, error, trace};
+
+pub mod config;
+pub mod enr;
+pub mod filter;
+pub mod metrics;
+
+pub use discv5::{self, IpMode};
+
+pub use config::{BootNode, Config, ConfigBuilder};
+pub use enr::uncompressed_id_from_enr_pk;
+pub use filter::{FilterOutcome, MustNotIncludeChains};
+use metrics::Metrics;
+
+/// Errors from using [`discv5::Discv5`] handle.
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// Failure adding node to [`discv5::Discv5`].
+    #[error("failed adding node to discv5, {0}")]
+    AddNodeToDiscv5Failed(&'static str),
+    /// Missing key used to identify rlpx network.
+    #[error("fork missing on enr, 'eth' key missing")]
+    ForkMissing,
+    /// Failed to decode [`ForkId`] rlp value.
+    #[error("failed to decode fork id, 'eth': {0:?}")]
+    ForkIdDecodeError(#[from] alloy_rlp::Error),
+    /// Peer is unreachable over discovery.
+    #[error("discovery socket missing")]
+    UnreachableDiscovery,
+    /// Peer is unreachable over rlpx.
+    #[error("rlpx TCP socket missing")]
+    UnreachableRlpx,
+    /// Peer is not using same IP version as local node in discovery.
+    #[error("discovery socket is unsupported IP version, local ip mode: {0:?}")]
+    IpVersionMismatchDiscovery(IpMode),
+    /// Peer is not using same IP version as local node in rlpx.
+    #[error("rlpx TCP socket is unsupported IP version, local ip mode: {0:?}")]
+    IpVersionMismatchRlpx(IpMode),
+    /// Failed to initialize [`discv5::Discv5`].
+    #[error("init failed, {0}")]
+    InitFailure(&'static str),
+    /// An error from underlying [`discv5::Discv5`] node.
+    #[error("{0}")]
+    Discv5Error(discv5::Error),
+    /// An error from underlying [`discv5::Discv5`] node.
+    #[error("{0}")]
+    Discv5ErrorStr(&'static str),
+}
+
+/// Use API of [`discv5::Discv5`].
+pub trait HandleDiscv5 {
```

Comment:
also not confident in this trait, these could be public methods on a struct and is implemented only once

Reply:
![we-hates-it](https://github.com/paradigmxyz/reth/assets/58548332/418a2f63-3fdc-41cf-9193-65e40ebcd4b8)